File: curbcheck/geocode/candidates.py
"""What a geocode answer is, and what each rung of the ladder is worth.

`suggest` and `reverse_geocode` both hand back one of these two records, and
every rung that can produce one scores it against the confidence constants
here, so the ladder is readable in one place rather than spread over the
modules that implement its rungs.
"""

from __future__ import annotations

from dataclasses import dataclass
from enum import StrEnum

from curbcheck.engine.coverage import COVERAGE_AREA

# Eight is the longest list a person scans without reading it as a search
# result page rather than as a disambiguation (docs/ux/UX_AUDIT.md P1-5 asks
# for candidates to always be offered, which only works if the list stays
# glanceable).
MAX_CANDIDATES = 8

# Everything in the database is in one borough, so the second line of a
# candidate says so when it has nothing more specific to say.
DEFAULT_SECONDARY = COVERAGE_AREA

CONFIDENCE_ADDRESS_POINT = 0.98
CONFIDENCE_INTERSECTION = 0.95
CONFIDENCE_PLACE = 0.85
CONFIDENCE_ADDRESS_INTERPOLATED = 0.75
CONFIDENCE_NEAR_ADDRESS = 0.60
CONFIDENCE_ADDRESS_RANGE = 0.50
# A street whose spelling the query matches exactly ("5 ave", "broadway") is
# what the user is naming, and has to outrank a place whose name merely
# contains those words: at 0.45 the query "5 ave" answered 5 AVE SYNAGOGUE
# (0.85 x 2 of its 3 words) before Fifth Avenue. A street reached by prefix
# ("broadwa") is still a guess about what is being typed, and one offered
# because half of an unmatched intersection hit it is a consolation prize that
# ranks below a partial place-name match.
CONFIDENCE_STREET_EXACT = 0.70
CONFIDENCE_STREET_WHOLE_QUERY = 0.45
CONFIDENCE_STREET_HALF_QUERY = 0.30
CONFIDENCE_ZIP = 0.25

# A place name matching only some of its own words is a weak hit ("86 ST"
# matches "CTL PK W DR OV 86 ST TRNVS RD"), so the score scales with how much
# of the place's name the query accounted for, and weak hits are dropped.
MIN_PLACE_COVERAGE = 0.5

# A fuzzy street match is a guess about what the user meant, so it costs a
# fifth of the candidate's confidence rather than being silently as good.
FUZZY_CONFIDENCE_PENALTY = 0.8


class GeocodeKind(StrEnum):
    """What a candidate *is*, which is what decides the icon and the second line.

    PIN is the one value the geocoder never returns: it is what the frontend
    labels a crosshair the user has not dropped yet. Everything else here is
    produced by `suggest` or `reverse_geocode`.
    """

    ADDRESS = "address"
    INTERSECTION = "intersection"
    STREET = "street"
    ZIP = "zip"
    PLACE = "place"
    PIN = "pin"


# A street's pin is a vertex of one of its own centerline segments and a corner
# is a segment endpoint (`etl.addresses._street_points`, `_write_intersections`),
# so both are at distance zero from the centerline by construction and asking
# `within_coverage` costs a scan to learn nothing. The kinds read off another
# dataset -- a surveyed door, a place, a ZIP centroid -- are still checked. On


@dataclass(frozen=True)
class GeocodeCandidate:
    """One place the query might mean. `confidence` is 0-1, higher is better.

    `label` is the line the user reads and `secondary` the muted line under it:
    the ZIP, how the point was arrived at, or the borough when there is nothing
    narrower to say. Both are the city's own capitals and are untrusted text.
    """

    label: str
    lat: float
    lon: float
    kind: GeocodeKind
    confidence: float
    secondary: str | None = DEFAULT_SECONDARY


@dataclass(frozen=True)
class ReverseMatch:
    """What a dropped pin is nearest to. `distance_m` is to the returned point."""

    label: str
    secondary: str | None
    kind: GeocodeKind
    lat: float
    lon: float
    distance_m: float


def zip_secondary(zipcode: object) -> str:
    text = str(zipcode or "").strip()
    return f"{DEFAULT_SECONDARY} {text}" if text else DEFAULT_SECONDARY

File: site/static/geocode/candidates.js
/**
 * What a geocode answer is, and what each rung of the ladder is worth.
 *
 * `suggest` and `reverseGeocode` both hand back one of these two records, and
 * every rung that can produce one scores it against the confidence constants
 * here, so the ladder is readable in one place rather than spread over the
 * modules that implement its rungs.
 *
 * The two records are the API payloads (`api/routes._candidate_payload`,
 * `_reverse_payload`): the field names are the Python dataclasses' own, and
 * `kind` is the string value of `GeocodeKind` rather than an enum member, so a
 * record can be posted back from the worker without a conversion step.
 */

import { COVERAGE_AREA } from "../engine/coverage.js";
import { NameMatch } from "./names.js";
import { pyStrip } from "./normalize.js";

// Eight is the longest list a person scans without reading it as a search
// result page rather than as a disambiguation (docs/ux/UX_AUDIT.md P1-5 asks
// for candidates to always be offered, which only works if the list stays
// glanceable).
export const MAX_CANDIDATES = 8;

// Everything in the database is in one borough, so the second line of a
// candidate says so when it has nothing more specific to say.
export const DEFAULT_SECONDARY = COVERAGE_AREA;

export const CONFIDENCE_ADDRESS_POINT = 0.98;
export const CONFIDENCE_INTERSECTION = 0.95;
export const CONFIDENCE_PLACE = 0.85;
export const CONFIDENCE_ADDRESS_INTERPOLATED = 0.75;
export const CONFIDENCE_NEAR_ADDRESS = 0.6;
export const CONFIDENCE_ADDRESS_RANGE = 0.5;
// A street whose spelling the query matches exactly ("5 ave", "broadway") is
// what the user is naming, and has to outrank a place whose name merely
// contains those words: at 0.45 the query "5 ave" answered 5 AVE SYNAGOGUE
// before Fifth Avenue. A street reached by prefix ("broadwa") is still a guess
// about what is being typed, and one offered because half of an unmatched
// intersection hit it is a consolation prize that ranks below a partial
// place-name match.
export const CONFIDENCE_STREET_EXACT = 0.7;
export const CONFIDENCE_STREET_WHOLE_QUERY = 0.45;
export const CONFIDENCE_STREET_HALF_QUERY = 0.3;
export const CONFIDENCE_ZIP = 0.25;

// What a word match is worth, by where the typed words landed in the name
// (`geocode/names.js`). Two bands, and the street band sits above the place
// band at every rung on purpose: a street is a destination this app can search
// both sides of for its whole length, a place is one point, so when the same
// word names both the street is the safer answer — "lex" is Lexington Avenue,
// not the Lex Hotel. The bands stay under CONFIDENCE_STREET_EXACT, so a street
// the user spelled out whole still beats every name merely containing that
// word; the only exception is a place named exactly what was typed, which
// cannot collide with a street because `etl.addresses.stage_places` drops a
// place whose name repeats a spelling of one.
export const STREET_NAME_CONFIDENCE = {
  [NameMatch.WHOLE]: CONFIDENCE_STREET_EXACT,
  [NameMatch.PREFIX]: 0.69,
  [NameMatch.LEADING]: 0.68,
  [NameMatch.INNER]: 0.67,
};
export const PLACE_NAME_CONFIDENCE = {
  [NameMatch.WHOLE]: CONFIDENCE_PLACE,
  [NameMatch.PREFIX]: 0.66,
  [NameMatch.LEADING]: 0.65,
  [NameMatch.INNER]: 0.64,
};

// A fuzzy street match is a guess about what the user meant, so it costs a
// fifth of the candidate's confidence rather than being silently as good.
export const FUZZY_CONFIDENCE_PENALTY = 0.8;

/**
 * What a candidate *is*, which is what decides the icon and the second line.
 *
 * PIN is the one value the geocoder never returns: it is what the frontend
 * labels a crosshair the user has not dropped yet. Everything else here is
 * produced by `suggest` or `reverseGeocode`.
 */
export const GeocodeKind = Object.freeze({
  ADDRESS: "address",
  INTERSECTION: "intersection",
  STREET: "street",
  ZIP: "zip",
  PLACE: "place",
  PIN: "pin",
});

/**
 * @typedef {object} GeocodeCandidate One place the query might mean.
 * @property {string} label the line the user reads; the city's own capitals, untrusted
 * @property {number} lat
 * @property {number} lon
 * @property {string} kind a `GeocodeKind` value
 * @property {number} confidence 0-1, higher is better
 * @property {string | null} secondary the muted line under it: the ZIP, how the
 *   point was arrived at, or the borough when there is nothing narrower to say
 */

/** @returns {GeocodeCandidate} */
export function geocodeCandidate({
  label,
  lat,
  lon,
  kind,
  confidence,
  secondary = DEFAULT_SECONDARY,
}) {
  return { label, lat, lon, kind, confidence, secondary };
}

/**
 * @typedef {object} ReverseMatch What a dropped pin is nearest to.
 * @property {string} label
 * @property {string | null} secondary
 * @property {string} kind a `GeocodeKind` value
 * @property {number} lat
 * @property {number} lon
 * @property {number} distance_m to the returned point
 */

/** @returns {ReverseMatch} */
export function reverseMatch({ label, secondary, kind, lat, lon, distanceM }) {
  return { label, secondary, kind, lat, lon, distance_m: distanceM };
}

export function zipSecondary(zipcode) {
  // `str(zipcode or "")`: a null, an empty string and a 0 all read as "no ZIP".
  const text = pyStrip(String(zipcode || ""));
  return text ? `${DEFAULT_SECONDARY} ${text}` : DEFAULT_SECONDARY;
}
